File: answers/leetcode/Count Complete Tree Nodes/Count Complete Tree Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftHeight = this.getHeight(root.left);
        int rightHeight = this.getHeight(root.right);
        int count = 1;
        if (leftHeight == rightHeight) {
            count += this.countNodes(root.right);
            for (int i = 0; i < leftHeight; i++) {
                count += 1 << i;
            }
        } else {
            count += this.countNodes(root.left);
            for (int i = 0; i < rightHeight; i++) {
                count += 1 << i;
            }
        }
        //System.out.println(root.val + ": " + leftHeight + " " + rightHeight + " " + count);
        return count;
    }
    
    private int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + this.getHeight(root.left);
    }
}

File: answers/leetcode/Count Complete Tree Nodes/Count Complete Tree Nodes.cpp
//@type Tree Binary Search
//@time complexity O(h ^ 2)
//@space complexity O(h)
//@result 18 / 18 test cases passed. Status: Accepted Runtime: 292 ms Submitted: 0 minutes ago You are here! Your runtime beats 35.69% of cpp submissions.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (NULL == root) {
			return 0;
		}
		int left(getLeft(root->left));
		if (getRight(root->right) == left) {
			return (2 << left) - 1;
		}
		else {
			return countNodes(root->left) + countNodes(root->right) + 1;
		}
    }
	int getLeft(TreeNode *root) {
		return NULL == root ? 0 : getLeft(root->left) + 1;
	}
	int getRight(TreeNode *root) {
		return NULL == root ? 0 : getRight(root->right) + 1;
	}
};
